x-pack/legacy/plugins/infra/public/store/v2/log_entries/index.ts

```diff
@@ -0,0 +1,240 @@
+  ReceiveEntriesBefore,
+  ReceiveEntriesAfter,
+  ErrorOnNewEntries,
+  ErrorOnMoreEntries,
+}
+
+type ReceiveActions =
+  | Action.ReceiveNewEntries
+  | Action.ReceiveEntriesBefore
+  | Action.ReceiveEntriesAfter;
+
+interface ReceiveEntriesAction {
+  type: ReceiveActions;
+  payload: LogEntriesState;
+}
+interface FetchOrErrorAction {
+  type: Exclude<Action, ReceiveActions>;
+}
+type ActionObj = ReceiveEntriesAction | FetchOrErrorAction;
+
+type Dispatch = (action: ActionObj) => void;
+
+interface LogEntriesDependencies {
+  filterQuery: SerializedFilterQuery | null;
+  timeKey: TimeKey | null;
+  pagesBeforeStart: number | null;
+  pagesAfterEnd: number | null;
+}
+
+const logEntriesInitialDependencies: LogEntriesDependencies = {
+  filterQuery: null,
+  timeKey: null,
+  pagesBeforeStart: null,
+  pagesAfterEnd: null,
+};
+
+type FetchEntriesParams = LogEntriesDependencies & {
+  sourceId: string;
+};
+
+interface FetchMoreEntriesParams {
+  pagesBeforeStart: number | null;
+  pagesAfterEnd: number | null;
+}
+
+export interface LogEntriesState {
+  entries: InfraLogEntry[];
+  entriesStart: TimeKey | null;
+  entriesEnd: TimeKey | null;
+  hasMoreAfterEnd: boolean;
+  hasMoreBeforeStart: boolean;
+  isReloading: boolean;
+  isLoadingMore: boolean;
+  lastLoadedTime: Date | null;
+}
+
+export interface LogEntriesCallbacks {
+  fetchNewerEntries: () => Promise<void>;
+}
+export const logEntriesInitialCallbacks = {
+  fetchNewerEntries: async () => {},
+};
+
+export const logEntriesInitialState: LogEntriesState = {
+  entries: [],
+  entriesStart: null,
+  entriesEnd: null,
+  hasMoreAfterEnd: false,
+  hasMoreBeforeStart: false,
+  isReloading: true,
+  isLoadingMore: false,
+  lastLoadedTime: null,
+};
+
+const shouldFetchNewEntries = ({
+  prevParams,
+  timeKey,
+  filterQuery,
+  entriesStart,
+  entriesEnd,
+}: FetchEntriesParams & LogEntriesState & { prevParams: FetchEntriesParams }) => {
+  if (!timeKey) return false;
+  const shouldLoadWithNewFilter = filterQuery !== prevParams.filterQuery;
+  const shouldLoadAroundNewPosition =
+    !entriesStart || !entriesEnd || !timeKeyIsBetween(entriesStart, entriesEnd, timeKey);
+  return shouldLoadWithNewFilter || shouldLoadAroundNewPosition;
+};
+
+enum ShouldFetchMoreEntries {
+  Before,
+  After,
+}
+
+const shouldFetchMoreEntries = ({ pagesAfterEnd, pagesBeforeStart }: FetchMoreEntriesParams) => {
+  if (pagesBeforeStart === null || pagesAfterEnd === null) return false;
+  if (pagesBeforeStart < DESIRED_BUFFER_PAGES) return ShouldFetchMoreEntries.Before;
+  if (pagesAfterEnd < DESIRED_BUFFER_PAGES) return ShouldFetchMoreEntries.After;
+  return false;
+};
+
+const useFetchEntriesEffect = (
+  dispatch: Dispatch,
+  state: LogEntriesState,
+  params: FetchEntriesParams
+) => {
+  const { getLogEntriesAround, getLogEntriesBefore, getLogEntriesAfter } = useGraphQLQueries();
+
+  const [prevParams, cachePrevParams] = useState(params);
+
+  const runFetchNewEntriesRequest = async () => {
```

Comment:
I think this seemed like an unnecessary optimization because I believe the effect only ever re-runs when `params` change, and all of these functions depend on `params`. I can wrap them in `useCallback`s just in case there's something I'm missing, though.